walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,183 @@
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<TimeRange>();
+    }
+    
+    // Build list of events that conflict with the mandatory requested attendees.
+    ArrayList<Event> attendedEvents = 
+        getEventsWithAttendees(events, request.getAttendees());
+
+    // Build list of events that conflict only with the optional attendees.
+    ArrayList<Event> attendedOptionalEvents = 
+        getEventsWithAttendees(events, request.getOptionalAttendees());
+
+    // Return with the full day if none of the events conflict with the requested attendees.
+    if (attendedEvents.isEmpty() && attendedOptionalEvents.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Separately get the time ranges of events on the calendars of the mandatory and 
+    // optional attendees.
+    ArrayList<TimeRange> eventTimeRanges = getEventTimeRanges(attendedEvents);
+    ArrayList<TimeRange> optionalEventTimeRanges = 
+                                           getEventTimeRanges(attendedOptionalEvents);
+
+    // If we only have mandatory attendees or only have optional attendees, find and 
+    // return that group's available time ranges.
+    if (eventTimeRanges.isEmpty()) {
+      return getAvailableTimeRanges(optionalEventTimeRanges, request.getDuration());
+    }
+
+    ArrayList<TimeRange> availableTimeRanges = 
+        getAvailableTimeRanges(eventTimeRanges, request.getDuration());
+    
+    if (optionalEventTimeRanges.isEmpty()) {
+      return availableTimeRanges;
+    }
+
+    // If we have both mandatory and optional attendees, combine the events so that
+    // optional events are only retained if they don't overlap with mandatory events.
+    // Find and return these time ranges only if there are still time ranges left.
+    combineTimeRanges(eventTimeRanges, optionalEventTimeRanges);
+    ArrayList<TimeRange> availableCombinedTimeRanges = 
+        getAvailableTimeRanges(eventTimeRanges, request.getDuration());
+    
+    if (availableCombinedTimeRanges.isEmpty()) {
+      // Adding optional events got rid of all possible time slots, so we want to 
+      // return the available time ranges from before we combined mandatory/optional.
+      return availableTimeRanges;
+    }
+
+    return availableCombinedTimeRanges;
+  }
+
+  /**
+   * Build a list of events that contain the given attendees.
+   */
+  private ArrayList<Event> getEventsWithAttendees(Collection<Event> events,
+                                                  Collection<String> attendees) {
+    ArrayList<Event> attendedEvents = new ArrayList<Event>();
+    for (Event event : events) {
+      for (String person : event.getAttendees()) {
+        if (attendees.contains(person)) {
+          attendedEvents.add(event);
+          break;
+        }
+      }
+    }
+
+    return attendedEvents;
+  }
+
+  /**
+   * Given a list of events, get the time ranges that the events take up, merging any 
```

Comment:
Likewise here, please also use the "summary fragment" style, e.g. "Gets the time ranges..." or "Returns the time ranges..."